src/main/java/uk/gov/justice/probation/courtcasematcher/restclient/OffenderSearchRestClient.java

```diff
@@ -1,64 +1,77 @@
 import org.springframework.stereotype.Component;
 import org.springframework.web.reactive.function.BodyInserters;
 import org.springframework.web.reactive.function.client.WebClient;
+import org.springframework.web.reactive.function.client.WebClientResponseException;
+import reactor.core.Exceptions;
 import reactor.core.publisher.Mono;
 import reactor.util.StringUtils;
-import uk.gov.justice.probation.courtcasematcher.event.OffenderSearchFailureEvent;
+import reactor.util.retry.Retry;
+import reactor.util.retry.Retry.RetrySignal;
 import uk.gov.justice.probation.courtcasematcher.model.externaldocumentrequest.Name;
 import uk.gov.justice.probation.courtcasematcher.model.offendersearch.MatchRequest;
 import uk.gov.justice.probation.courtcasematcher.model.offendersearch.SearchResponse;
 
 import static org.springframework.security.oauth2.client.web.reactive.function.client.ServletOAuth2AuthorizedClientExchangeFilterFunction.clientRegistrationId;
 
-@SuppressWarnings("UnstableApiUsage")
 @Component
 @Slf4j
 public class OffenderSearchRestClient {
     private static final String ERROR_NO_DATE_OF_BIRTH = "No dateOfBirth provided";
     private static final String ERROR_NO_NAME = "No surname provided";
+
+    @Setter
     @Value("${offender-search.post-match-url}")
     private String postMatchUrl;
 
+    @Setter
     @Value("${offender-search.disable-authentication:false}")
     private Boolean disableAuthentication;
 
-    private final WebClient webClient;
+    @Setter
+    @Value("${offender-search.max-retries:3}")
+    private int maxRetries;
 
-    private final EventBus eventBus;
+    @Setter
+    @Value("${offender-search.min-backoff-seconds:5}")
+    private int minBackOffSeconds;
 
-    private final ObjectMapper objectMapper;
+    private final WebClient webClient;
 
     @Autowired
-    public OffenderSearchRestClient(@Qualifier("offenderSearchWebClient") WebClient webClient, EventBus eventBus, com.fasterxml.jackson.databind.ObjectMapper objectMapper) {
+    public OffenderSearchRestClient(@Qualifier("offenderSearchWebClient") WebClient webClient) {
         super();
         this.webClient = webClient;
-        this.eventBus = eventBus;
-        this.objectMapper = objectMapper;
     }
 
     public Mono<SearchResponse> search(Name name, LocalDate dateOfBirth){
         if (!validate(name, dateOfBirth)) {
-            return Mono.empty();
+            return Mono.error(new IllegalArgumentException("Invalid parameters passed for offender search"));
         }
         MatchRequest body = buildRequestBody(name, dateOfBirth);
         return post()
                 .body(BodyInserters.fromPublisher(Mono.just(body), MatchRequest.class))
                 .accept(MediaType.APPLICATION_JSON)
                 .retrieve()
                 .bodyToMono(SearchResponse.class)
-                .onErrorResume(throwable -> handleError(throwable, body));
+                .retryWhen(Retry.backoff(maxRetries, Duration.ofSeconds(minBackOffSeconds))
```

Comment:
This is really quite neat and reads really well given how much this method is doing at this point. It feels like WebClient is beginning to pay dividends in readability here.

Reply:
Yes - agreed. I am going to raise a bug for retry on the PUT of court case because we had some timeouts on that this morning. Now we have one, should be fairly simple to add others